### ISSUE TITLE
Frequent test failures since 0.15

### ISSUE THREAD

**Issue by kazuho** [kazuho/p5-Server-Starter#13]
http://matrix.cpantesters.org/?dist=Server-Starter-0.14
http://matrix.cpantesters.org/?dist=Server-Starter-0.15
http://matrix.cpantesters.org/?dist=Server-Starter-0.16
http://matrix.cpantesters.org/?dist=Server-Starter-0.17

**Comment by kazuho** [comment]
Might not be a timing issue.

Installed vanilla NetBSD 6.1.5 (amd64) onto VMware, and t/06-autorestart.t always fail (6 out of 6 tries).

The log is https://gist.github.com/kazuho/4f9106924f03c3428dd5, perl version is 5.20.1 (installed using pkgsrc).

**Comment by kazuho** [comment]
Well, it was actually a timing issue.

The timings tests were in fact too fragile since `sleep(n)` gets interrupted by child exists (since SIGCHLD is delivered).

Many of the fixes being pushed have their timing logs attached as gists.

**Comment by kazuho** [comment]
To tackle the issue I have done the following:

- reimplement the changes in 0.15, 0.16
 - includes the removal of `sleep(1)` from the main loop
  - also no more causes unnecessary wakeups unless `--auto-restart` is used
- adjust the timings by looking at the timing logs on OSX and NetBSD

I am going to make a dev release (0.18_02) so that we could see the results on cpan test matrix.

**Comment by limitusus** [comment]
Looks clear :+1: 
I'll check CPAN test matrix later too.

**Comment by kazuho** [comment]
Thank you for looking into the changes.  The test matrix seem to have improved http://matrix.cpantesters.org/?dist=Server-Starter+0.18_02

I have released 0.19 with the changes.  Please let me know if you see any problem.

**Comment by limitusus** [comment]
:ok_hand:

**Comment by ppisar** [comment]
If you incorporated my patches <https://rt.cpan.org/Public/Bug/Display.html?id=73711> which you ignored, you could get more reliable tests. The key fix is stop using magical sleeps because you never can be sure how operating systems scheduler plans the CPU time.

**Comment by kazuho** [comment]
@ppisar Sorry for not responding to the RT.  I rarely visit them.

I agree with the fact that relying on sleep is a bad thing, but in this case the tests need to assert that the command is actually calling sleeps before performing the next action.  Therefore they are necessary.

**Comment by ppisar** [comment]
On Mon, May 25, 2015 at 01:51:04PM -0700, Kazuho Oku wrote:
> @ppisar Sorry for not responding to the RT.  I rarely visit them.
> 
> I agree with the fact that relying on sleep is a bad thing, but in this case
> the tests need to assert that the command is actually calling sleeps before
> performing the next action.  Therefore they are necessary.
> 
Where the sleep is necessary, you have to meassure elapsed real time and skip
affected tests if the time is larger than expected. (For example test for
restarted server by matching /^2:/ will probably fail if more than two restart
cycles pass before parsing the output.

-- Petr

**Comment by kazuho** [comment]
@ppisar Thank you for the suggestion.

> Where the sleep is necessary, you have to meassure elapsed real time and skip affected tests if the time is larger than expected.

Unfortunately such a workaround is unlikely to work correctly, since the code being tested run on another process, which are also scheduled by the OS.

On a server with such a high load (where returning from a `sleep` gets delayed for a substantial time) it is unlikely that other processes (which do heavier things like fork and exec) will run in expected time.

In other words, there is a requirement that the tests should not be executed on a heavily-loaded host.  Theoretically the requirement might cause false-positives, but IMO up until now running _more_ tests on various environments has helped improve the quality of the software more than the false-positives being reported.

**Comment by ppisar** [comment]
On Tue, May 26, 2015 at 02:49:12PM -0700, Kazuho Oku wrote:
> @ppisar Thank you for the suggestion.
> 
> > Where the sleep is necessary, you have to meassure elapsed real time and
> > skip affected tests if the time is larger than expected.
> 
> Unfortunately such a workaround is unlikely to work correctly, since the
> code being tested run on another process, which are also scheduled by the
> OS.
> 
Fortunately such workaround is likely to work correctly. Much more likely
than current tests. This is my experience. I think this is because operating
system distributes processess to CPU's evenly, so if one CPU is loaded, very
likely all the CPUs are loaded.

It also points to lack of missing synchronization between the tests and another
process. For example atomic replacement of the status file generated by the
worker or waiting for one-line status file by the tests which I also
implemented in the patches fixes many of the races.

Definitely the less races are there the more reliable the tests are.

> On a server with such a high load (where returning from a `sleep` gets
> delayed for a substantial time) it is unlikely that other processes (which
> do heavier things like fork and exec) will run in expected time.
> 
> In other words, there is a requirement that the tests should not be executed
> on a heavily-loaded host.  Theoretically the requirement might cause
> false-positives, but IMO up until now running _more_ tests on various
> environments has helped improve the quality of the software more than the
> false-positives being reported.
> 
Unfortunately for me, current code fails almost in quater of the runs. When
using my patches, I do not notice any false negatives.

-- Petr

**Comment by kazuho** [comment]
@ppisar Can you point out any timing issues in http://matrix.cpantesters.org/?dist=Server-Starter+0.27?  Most of them are passing, and the failing ones do not look like timing issues (at least the ones I've seen are).  Are you using the newest version of Server-Starter, or are you referring to the issue that was fixed by this PR?

And regarding the topic of the discussion, I think you misunderstand the point.  Let me rephrase:

Server::Starter is expected to insert some delay (specified by the user) between launching processes.  I am against adding complexity to the library just for _theoretically_ improving the stability of tests under the assumption that current tests are stable enough.

In other words, no synchronization exists between Server::Starter (the process being tested) and the tests being executed.  So it is impossible to provide a test that would run on a highly-loaded host in which `sleep` will consume much longer time than requested.

If you are seeing issues with the current version, please let me know the details of the failures and I am willing to discuss how we should fix them.

**Comment by ppisar** [comment]
On Wed, May 27, 2015 at 12:16:31AM -0700, Kazuho Oku wrote:
> @ppisar Can you point out any timing issues in
> http://matrix.cpantesters.org/?dist=Server-Starter+0.27?

This one <https://bugzilla.redhat.com/show_bug.cgi?id=1224727>. Obviously the
"sleep 3;" was not enough in the case:

    like(get_status(), qr/^1:\d+\n2:\d+$/s, 'status during transient state');
    sleep 3;
    like(get_status(), qr/^2:\d+\n$/s, 'status after auto-restart');

The test catches the status file at a time when it still contains two lines.
One from old and one from new worker, thus t/06-autorestart.t test expecting
/^2:\d+\n$/ fails.

> Are you using the newest version of Server-Starter,
Yes, the 0.27.

> Server::Starter is expected to insert some delay (specified by the user)
> between launching processes.  I am against adding complexity to the library
> just for _theoretically_ improving the stability of tests under the
> assumption that current tests are stable enough.
> 
> In other words, no synchronization exists between Server::Starter (the
> process being tested) and the tests being executed.  So it is impossible to
> provide a test that would run on a highly-loaded host in which `sleep` will
> consume much longer time than requested.
>
I undestand you do not want to add complexity to the library. Therefore
I proposed adding complexity to tests which significantly reduces false
failures.

For example changing get_status() to block until exactly one line exist in the
file helps.

Another improvement is to remember recently retrieved generation
number (the value before the colon) and wait until the generation number
increases.

Then the test will still correcly check that the the library restarts the
workers. It will not assure that it managed to do excact amount of restarts in
an exact time period, but as you wrote, it's not possible to create such test
working under all conditions without adding synchronization into the library.

(Actually even the library cannot guarantee the exact times for the very same
reason, so that implies there is no point in testing the exact times. In case
you would not like lossing the exact time tests, I proposed a work around: If
the test observers that running time exceeds expected time, then it will skip
the tests because the assumption of the same time refence was invalidated.)

-- Petr
